### DISCUSSION TITLE
Undo/Redo geometries in Mayo

### DISCUSSION THREAD

**Discussion by thyis** [fougue/mayo#239] [General]
Dear Sir:
First! Thanks for developing the wonderful cad viewer mayo, it it really a very good software!
If possible, could you please kindly help to provide some details regarding how geometries can be undo/ redo in mayo using occ interfaces? This would be helpful for some simple geometry  creatations and undo/redo (such as sphere etc).

**Comment by HuguesDelorme** [comment]
Basically, you would use the undo framework provided by `TDocStd_Document` which `Mayo::Document` inherits from.
Assuming you have some `Mayo::Document` object:
````c++
doc->NewCommand();
// Make some modifications to the document, eg import file, create primitive shapes, ...
doc->CommitCommand();
````

Undo the recorded changes:
````c++
if (doc->Undo())
    doc->CommitCommand();
````

Redo after call to Undo():
````c++
if (doc->Redo())
    doc->CommitCommand();
````

**Comment by thyis** [comment]
Thanks a lot! This is the exact information I was looking for.
